src/main/java/org/craftercms/engine/util/servlet/ConfigAwareProxyServlet.java

```diff
@@ -84,4 +111,155 @@ protected void copyResponseHeaders(HttpResponse proxyResponse, HttpServletReques
+        if (servletRequest.getAttribute(ATTR_TARGET_URI) == null) {
+            servletRequest.setAttribute(ATTR_TARGET_URI, targetUri);
+        }
+        if (servletRequest.getAttribute(ATTR_TARGET_HOST) == null) {
+            servletRequest.setAttribute(ATTR_TARGET_HOST, targetHost);
+        }
+        String proxyRequestUri = rewriteUrlFromRequest(servletRequest);
+        URL url = new URL(proxyRequestUri);
+
+        // Create a new socket connection to the websocket server
+        Socket socketProxyClient = new Socket(url.getHost(), url.getPort());
+        boolean closeSocket = false;
+        try {
+            var socketIn = socketProxyClient.getInputStream();
+            var socketOut = socketProxyClient.getOutputStream();
+
+            // Send the handshake request from engine to the websocket server
+            String handshakeRequest = getProxyHandshakeRequest(servletRequest, url);
+            socketOut.write(handshakeRequest.getBytes(StandardCharsets.UTF_8));
+            socketOut.flush();
+
+            // Get handshake response from websocket server to engine
+            String[] responseLines = getProxyHandshakeResponse(socketIn);
+            addProxyResponseHeaders(responseLines, servletResponse);
+            int respCode = proxyHandshakeResponseCode(responseLines);
+            if (respCode != HttpServletResponse.SC_SWITCHING_PROTOCOLS) {   // The websocket handshake failed, close the connection
+                servletResponse.setStatus(respCode);
+                servletResponse.flushBuffer();
+                logger.debug("< Websocket| Flush");
+                closeSocket = true;
+            } else { // The websocket handshake succeeded, establish the websocket connection
+                var upgradeHandler = servletRequest.upgrade(WsUpgradeHandler.class);
+                upgradeHandler.preInit(exec, socketIn, socketOut, socketProxyClient);
+            }
+        } finally {
+            if (closeSocket) {
+                socketProxyClient.close();
+            }
+        }
+    }
+
+    /**
+     * Get the handshake request to websocket server
+     * @param servletRequest original servlet request from client
+     * @param serverUrl websocket server url
+     * @return GET request for handshaking websocket connection
+     */
+    private String getProxyHandshakeRequest(HttpServletRequest servletRequest, URL serverUrl) {
+        StringBuilder req = new StringBuilder(512);
+        req.append("GET " + serverUrl.getFile()).append(" HTTP/1.1");
+        logger.debug("> Websocket|", req);
+        req.append("\r\n");
+        var headerNames = servletRequest.getHeaderNames();
+        while (headerNames.hasMoreElements()) {
+            var name = headerNames.nextElement();
+            String header = servletRequest.getHeader(name);
+            logger.debug("> Websocket| {}:{}", name, header);
+            req.append(name + ": " + header + "\r\n");
```

Comment:
Since this is in a loop, let's use a string formatter instead of `+` for performance.

Reply:
Fixed to use the `req` StringBuilder